syntheseus/interface/reaction.py

```diff
@@ -30,12 +29,11 @@ class ReactionMetaData(TypedDict, total=False):
 
 
 @dataclass(frozen=True, order=False)
-class Reaction(Generic[ReactantsType, ProductType]):
+class Reaction(Generic[ReactantsType]):
     """General reaction class."""
 
     # The molecule that the prediction is for and the predicted output:
     reactants: ReactantsType = field(hash=True, compare=True)
-    product: ProductType = field(hash=True, compare=True)
     identifier: Optional[str] = field(default=None, hash=True, compare=True)
```

Comment:
I think our reaction types hierarchy is a bit overcomplicated given what it tries to accomplish. What about having the `products` attribute in the base class, getting rid of `MultiProductReaction` entirely, and having `SingleProductReaction` add only a `@property` that asserts there is exactly one product and returns it (without adding any actual attributes)?

Reply:
Additionally, perhaps `Reaction` doesn't need to be generic at all; we may just want to fix the types of `reactants` and `products` to `Bag[Molecule]` (and the `product` property would return `Molecule`).

Reply:
I thought about that, but the disadvantage of this approach would be that construction is more awkward: you need would write `SingleProductReaction(products={mol}, ...)` (i.e. always wrapping mol in a set or Bag) and the restriction that there is only one mol would only be checked when you try to *access* it, which is not ideal in my opinion.

I do agree that it could be simplified though, for example by making the class no longer generic. What about just removing the generic part?

Reply:
For example, what about the following class structure @kmaziarz ?

```python
class _BaseReaction:
    reactants: Bag[Molecule]
    ... # identifier / metadata

class Reaction(_BaseReaction):
    products: Bag[Molecule]

class SingleProductReaction(_BaseReaction)
    product: Molecule
```

(note: `@dataclass` flags and string methods are not shown)

This is less complicated than my current solution, but still allows `SingleProductReaction` objects to be created without an empty `Bag` wrapper. We could even add a `products` property to `SingleProductReaction` if we wanted to.

Reply:
> what about the following class structure

This looks fine, it is a step in the right direction.

> I thought about that, but the disadvantage of this approach would be that construction is more awkward: you would write `SingleProductReaction(products={mol}, ...)` (i.e. always wrapping mol in a set or bag) and the restriction that there is only one mol would only be checked when you try to _access_ it, which is not ideal in my opinion.

Hmm, yes; the later could be tackled with a check in `__post_init__`, but the former may be tricky to get around.

Reply:
> This looks fine, it is a step in the right direction.

What would you propose instead? I agree it's not ideal but none of the options seem ideal...

Reply:
Yeah, I agree, none of what we discussed seems ideal. Here's another approach: what about overriding the auto-generated `__init__` in `SingleProductReaction`:

```python
@dataclass
class Reaction:
    reactants: Bag[Molecule]
    products: Bag[Molecule]


@dataclass
class SingleProductReaction(Reaction):
    def __init__(self, reactants: Bag[Molecule], product: Molecule) -> None:
        super().__init__(reactants=reactants, products=Bag([product]))

    @property
    def product(self) -> Molecule:
        assert len(self.products) == 1  # Guaranteed in `__init__`.
        return self.products[0]
```
Wrapping the product into a `Bag` would only appear in one place (`__init__` in `SingleProductReaction`), and that also guarantees that `len(self.products) == 1` for any instance of `SingleProductReaction` (so there would be no way to make the `assert` fail without some low-level Python hackery).

Reply:
That approach makes sense, but is a bit complex. It almost just seems like a complicated way to ensure that `SingleProductReaction`s have a `products` attribute (with an s). Is that even a desirable outcome?

If it is just about the class hierarchy then we can just simply it using a `typing.Protocol`. Would that be better?

Reply:
> It almost just seems like a complicated way to ensure that `SingleProductReaction`s have a products attribute (with an s). Is that even a desirable outcome?

My thinking was that this would allow us to process both kinds of reactions in a unified way. Currently `eval_single_step.py` if-cases this; it would be nice if we could compare the produced reaction to the one from the dataset by just comparing `reactants` and `products` directly.